drivers: fpga: Added altera FPGA bridge support

hardeepsharma95 opened the pull request:
Added FPGA Bridge Enable/Disable functionality :

Adding support for Agilex & Agilex5

hardeepsharma95 committed:
drivers: fpga: Added altera FPGA bridge support

Added altera FPGA bridge support

Signed-off-by: Hardeep Sharma <hardeep.sharma@intel.com>

hardeepsharma95 committed:
dts: intel_socfpga: Added altera FPGA driver dtsi node

Added altera FPGA driver dtsi node

Signed-off-by: Hardeep Sharma <hardeep.sharma@intel.com>

hardeepsharma95 commented:
As discussed in the Architecture group meeting moved fpga bridges from subsystem to driver.
Old PR :: #65753 

Kindly review this PR

tgorochowik commented on drivers/fpga/fpga_altera.c:
```diff
@@ -0,0 +1,429 @@
+
+static int altera_fpga_on(const struct device *dev)
+{
+    int32_t ret = 0;
```
Incorrect indentation

tgorochowik commented on drivers/fpga/fpga_altera.c:
```diff
@@ -0,0 +1,429 @@
+
+static int altera_fpga_init(const struct device *dev)
+{
+    if (!dev) {
```
Same here

hardeepsharma95 replied:
done

tgorochowik commented:
Looks nice in general, please fix the compliance checks first (note you can also run them locally, so you don't have to wait for the CI run to be approved with each iteration)

tgorochowik commented on drivers/fpga/fpga_altera.c:
```diff
@@ -0,0 +1,429 @@
+    return 0;
+}
+
+static const struct fpga_driver_api altera_fpga_api = {
+	.on = altera_fpga_on,
+	.off = altera_fpga_off
+};
```
Unfortunately the subsystem currently requires the remaining callbacks to be defined, so that cannot go in like this.
At the very least we'd need to modify the common layer to allow NULLs for the rest and return -ENOTSUP. Otherwise you might see random issues when sb tries to use your driver

hardeepsharma95 replied:
done

tgorochowik commented on drivers/fpga/fpga_altera.c:
```diff
@@ -0,0 +1,463 @@
+static const char *altera_fpga_get_info(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return NULL;
```
I am afraid it still won't work, the FPGA layer is too limited at this point.

You'll get a NULL here: https://github.com/zephyrproject-rtos/zephyr/blob/9b17a952f6df812fad42e124e2b70eea89c77ff5/include/zephyr/drivers/fpga.h#L126-L132

and it will propagate here:
https://github.com/zephyrproject-rtos/zephyr/blob/9b17a952f6df812fad42e124e2b70eea89c77ff5/drivers/fpga/fpga_shell.c#L127

It is similar for other functions as well.

tanmaykathpalia replied:
@tgorochowik  Agree, NULL will be propagated to fpga_shell but we can't return -ENOSUP from the common layer i.e. fpga.h as the return type for the fpga_get_info function is of type "char *" or you are suggestion is to return "-ENOSUP" as string.
Either way, we have to return NULL or "-ENOSUP" either from the driver or from the common layer.

cfriedt replied:
Is there a significant reason to return `NULL` or `"-ENOTSUP"` here?

For lack of a better option, you could return e.g. the crc of the bitstream if it is loaded, otherwise, an empty string.

https://github.com/zephyrproject-rtos/zephyr/blob/e3ff9930003cfbbaf1409d4aee51e39057358b0d/drivers/fpga/fpga_ice40.c#L495

hardeepsharma95 replied:
Thanks @cfriedt for you valuable comment I agree with you we can return the empty string in this case

tgorochowik reviewed (changes requested):
If we want to allow having only selected capabilities, we need to expand the common layer imo, e.g, add checks for NULL here:
https://github.com/zephyrproject-rtos/zephyr/blob/9b17a952f6df812fad42e124e2b70eea89c77ff5/include/zephyr/drivers/fpga.h#L128-L131

And return -ENOSUP there (in the common layer).

In addition we need to make sure the returned value is used wherever the driver is used (e.g. in the fpga shell implementation).

Otherwise I think it will too fragile.

cfriedt commented on drivers/fpga/fpga_altera.c:
```diff
@@ -0,0 +1,463 @@
+}
+
+/* Will send the SMC command to check the status of the FPGA */
+static int32_t is_fpga_config_not_ready(void)
```
It might be better to rename this so that it provides non-inverted logical operation (adjust return value as necessary).

hardeepsharma95 replied:
done

cfriedt commented on drivers/fpga/fpga_altera.c:
```diff
@@ -0,0 +1,463 @@
+	}
+
+	/* Check FPGA status before bridge enable/disable */
+	ret = is_fpga_config_not_ready();
```
I would suggest using a name that implies non-inverted logic

hardeepsharma95 replied:
done

hardeepsharma95 commented on drivers/fpga/fpga_altera.c:
```diff
@@ -0,0 +1,463 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT intel_fpga
```
Thanks for your suggestion @cfriedt I made the required changes

cfriedt commented on drivers/fpga/fpga_altera.h:
```diff
@@ -0,0 +1,207 @@
+		 * is useful for matching responses to commands along with the ID
+		 */
+		uint32_t client_id : 4;
+	} header_t;
```
this is not a typedef so it probably makes sense to remove the `_t`

hardeepsharma95 replied:
done

cfriedt commented on drivers/fpga/fpga_altera.h:
```diff
@@ -0,0 +1,207 @@
+#define ZEPHYR_SUBSYS_FPGA_BRIDGE_INTEL_H_
+
+#include <zephyr/kernel.h>
+#include <zephyr/sip_svc/sip_svc.h>
+#include <zephyr/drivers/sip_svc/sip_svc_agilex_smc.h>
```
Are these two headers required by anything in this file? If not, please include them where they are needed.

hardeepsharma95 replied:
done

cfriedt commented on dts/arm64/intel/intel_socfpga_agilex.dtsi:
```diff
@@ -99,6 +99,10 @@
 		reg = <0x10000000 0x200000>;
 	};
 
+	fpga0: bridges {
```
Shouldn't there be an address associated with this device inside of the fpga SoC?

hardeepsharma95 replied:
Actaully we are not directly interacting with any hardware we are calling sip_svc subsystem which further request the ATF to read and write in the bridge register so this bridge dont have any address

cfriedt commented on dts/arm64/intel/intel_socfpga_agilex5.dtsi:
```diff
@@ -97,6 +97,10 @@
 		reg = <0x80100000 DT_SIZE_M(8)>;
 	};
 
+	fpga0: bridges {
```
Shouldn't there be an address associated with this device inside of the fpga SoC?

hardeepsharma95 replied:
Same for here

cfriedt commented on drivers/fpga/fpga_altera.h:
```diff
@@ -0,0 +1,207 @@
+};
+
+/* SIP SVC response private data */
+struct private_data_t {
```
This is not typedef'ed so please remove the `_t`
```suggestion
struct private_data {
```

hardeepsharma95 replied:
done

cfriedt commented on drivers/fpga/Kconfig.altera:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2024, Intel Corporation. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+config ALTERA_FPGA
```
Since this is only one type of interface (agilex bridge) it might be best to rename this option, because. Otherwise, it might be mistaken as support for all altera fpgas

cfriedt commented on drivers/fpga/fpga_altera.c:
```diff
@@ -0,0 +1,463 @@
+/*
```
File name should have agilex bridge appended

hardeepsharma95 replied:
done

tgorochowik commented on drivers/fpga/fpga_altera_agilex_bridge.c:
```diff
@@ -0,0 +1,464 @@
+
+static enum FPGA_status altera_fpga_get_status(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return -ENOTSUP;
```
You're adding on/off code above - can this function be implemented to check current on/off state?

tgorochowik replied:
Maybe not, I've proposed a separate PR which should make it simpler to contribute drivers that do not implement the whole API.

hardeepsharma95 replied:
Should I remove these un-used API for now from the code as anyway in your PR you are taking care for that part ?

tgorochowik replied:
I you do that then you'd have to wait for my PR to be merged first. I am ok with adding empty functions here for now in case my PR gets stuck for some reason (and then cleaning up your driver later)

hardeepsharma95 replied:
ok I will clean up later then . Can you please approve my PR for now ?

hardeepsharma95 commented:
Hello All , Kindly Review and Approve this PR

hardeepsharma95 commented:
Hello @tgorochowik @cfriedt , Kindly Review and Approve this PR

cfriedt commented:
#71500 might be better to merge first. Sorry for the delay. Then there is no reason to add a function that returns the empty string.

hardeepsharma95 commented:
> #71500 might be better to merge first. Sorry for the delay. Then there is no reason to add a function that returns the empty string.

@cfriedt is it okay to let this PR merge ? Later I will create another PR to clean up the code

hardeepsharma95 commented:
Hi @tgorochowik @cfriedt as this PR (https://github.com/zephyrproject-rtos/zephyr/pull/71500) got merged .

Kindly review and approve this PR

cfriedt commented on drivers/fpga/fpga_altera_agilex_bridge.c:
```diff
@@ -0,0 +1,433 @@
+
+LOG_MODULE_REGISTER(fpga_altera);
+
+static uint32_t mailbox_client_token;
+static struct sip_svc_controller *mailbox_smc_dev;
```
These are duplicates definitions and should be removed.

hardeepsharma95 replied:
done

cfriedt commented on drivers/fpga/fpga_altera_agilex_bridge.c:
```diff
@@ -0,0 +1,433 @@
+
+#define MAX_TIMEOUT_MSECS (1 * 1000UL)
+
+static uint32_t mailbox_client_token;
+static struct sip_svc_controller *mailbox_smc_dev;
```
The devicetree bindings for this driver should include a phandle property that refers to the sip_svc_controller.

That should be stored as compile-time const data in your `struct <driver>_config`.

Similarly, the mailbox client token should be part of a `struct <driver>_data` assuming it is mutable and only runtime discoverable.

tanmaykathpalia replied:
@cfriedt As discussed in the architecture review meeting, we have made a dummy node in the device tree for this driver as it doesn't need the information about the SiP SVC controller.  As the SiP SVC device node will be 
 returned by "sip_svc_get_controller" function declared in "zephyr/sip_svc/sip_svc.h" file and it is a subsystem, not a driver.

Yes, for the other comment regarding keeping the information in the struct <driver>_data, we agree it should be kept the way you mentioned.

cfriedt commented on drivers/fpga/fpga_altera_agilex_bridge.c:
```diff
@@ -0,0 +1,433 @@
+	.off = altera_fpga_off,
+};
+
+DEVICE_DT_DEFINE(DT_NODELABEL(fpga0), &altera_fpga_init, NULL, NULL, NULL, POST_KERNEL,
```
It would be preferable to instantiate this with a macro - `DT_INST_FOREACH_STATUS_OKAY()`.

hardeepsharma95 replied:
done

cfriedt reviewed (dismissed):
Looks OK for me, with the caveat that the Arch WG approved of having no DT entry. It would be nice to have eventually, but it's not a blocking issue.

cfriedt commented:
@hardeepsharma95 - I might have approved prematurely. We still need a "build_all" test that incorporates this driver into CI so that we know it's healthy.

hardeepsharma95 committed:
tests: drivers: build_all: Build Altera FPGA driver

Altera FPGA driver should be built on a regular basis to ensure
that there are no regressions

Signed-off-by: Hardeep Sharma <hardeep.sharma@intel.com>

hardeepsharma95 commented:
> @hardeepsharma95 - I might have approved prematurely. We still need a "build_all" test that incorporates this driver into CI so that we know it's healthy.

done

hardeepsharma95 commented on tests/drivers/build_all/fpga/boards/intel_socfpga_agilex5_socdk.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_ARM_SIP_SVC_DRIVER=y
+CONFIG_ARM_SIP_SVC_SUBSYS=y
+CONFIG_ARM_SIP_SVC_SUBSYS_SINGLY_OPEN=y
+CONFIG_HEAP_MEM_POOL_SIZE=16384
```
done

hardeepsharma95 commented on tests/drivers/build_all/fpga/boards/intel_socfpga_agilex_socdk.conf:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2024, Intel Corporation.
```
done

hardeepsharma95 commented:
Hi @tgorochowik @cfriedt

Kindly review and approve this PR

hardeepsharma95 commented:
Hi @tgorochowik @cfriedt

Kindly review and approve this PR

cfriedt reviewed (approved):
LGTM - thanks for your patience 🙏

hardeepsharma95 commented:
Hello All  ,
Kindly review and approve this PR
